PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+#pragma once
+
+#include "BTreeNode.hpp"
+#include "Optional.hpp"
+
+namespace Poly {
+
+	template<typename K, typename V, size_t B>
+	struct Range { //note(vuko): doubles as a proper fucking iterator; behold the goddamn beauty! (alright, this is for show only for now)
+		using BTree  = Impl::BTree<K, V, B>;
+		using KVERef = typename BTree::KVERef;
+
+		KVERef front;
+		KVERef back;
+
+		struct KV {
+			const K& key;
+			V& value;
+		};
+
+		KV next_unchecked() {
+			KVERef kv_ref = this->front;
+
+			if (kv_ref.idx < kv_ref.node_ref.node->len) {
+				front = KVERef{kv_ref.node_ref, kv_ref.idx + 1};
+				return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+			}
+
+			do {
+				kv_ref = kv_ref.node_ref.ascend().parent;
+			} while (kv_ref.idx >= kv_ref.node_ref.node->len);
+
+			front = BTree::first_leaf_edge(KVERef{kv_ref.node_ref, kv_ref.idx + 1}.descend());
+			return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+		}
+
+		KV prev_unchecked() {
+			KVERef kv_ref = this->back;
+
+			if (kv_ref.idx > 0) {
+				this->back = KVERef{kv_ref.node_ref, kv_ref.idx - 1};
+				return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+			}
+
+			do {
+				kv_ref = kv_ref.node_ref.ascend().parent;
+			} while (kv_ref.idx == 0);
+
+			this->back = BTree::last_leaf_edge(KVERef{kv_ref.node_ref, kv_ref.idx - 1}.descend());
+			return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+		}
+
+		KV next() {
+			assert(!(front == back));
+			return next_unchecked();
+		}
+	};
```

Comment:
I'd rather have it in a separate PR. Doing it that way will allow me to iron out the kinks in design and won't delay the map any longer. As a side note, a specialized map range can be implemented way more efficiently than the generic one implemented in terms of iterators. In fact for B-trees the iterators themselves are more elegantly implemented in terms of a range encompassing the whole collection and not the other way around!